### PR TITLE
DM-4708:  Integrate astrometry test into SDSS demo lsst_dm_stack_demo

### PR DESCRIPTION
**Pull request by wmwv**
Add the option of calculating the astrometry RMS for the LSST stack demo of SDSS data.

After running the basic `./bin/demo.sh`, you can then run

```
python bin/check_astrometry.py output
```

to generate the output for the astrometry RMS calculation.

This has not been added to the bin/compare script due to the expectation that the astrometric RMS will vary at significantly greater than numerical precision for different reasonable choices of astrometric solution.

### CODE REVIEW

File: README [hunk 1]
```diff
@@ -42,6 +42,10 @@ If you use the "--small" run option then the corresponding output file is
 named detected-sources_small.txt and the corresponding output directory is
 called "output_small".
 
+Check the astrometric relative RMS with
```
Comment (jdswinbank) [inline README]: This looks like it's supposed to be ReST, so please finish the text preceding a block of code with `::` (see elsewhere in the file for examples).

File: bin/check_astrometry.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
+#!/usr/bin/env python2
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: I'd much rather see `/usr/bin/env python`.

Also, there should be a copyright blurb in here.

Also, please `from __future__ import print_function` and change the print statements below appropriately.

File: bin/check_astrometry.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+import math
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: `math` isn't used.

File: bin/check_astrometry.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+import matplotlib.pylab as plt
+import numpy as np
+
+import lsst.daf.persistence as dafPersist
+import lsst.pipe.base as pipeBase
+import lsst.afw.table as afwTable
+import lsst.afw.geom as afwGeom
+import lsst.afw.image as afwImage
+
+def loadData(repo, visits, field, ref, ref_field, camcol, filter) :
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: It seems like this function is doing a lot more than just loading data!

At a minimum, could you give it a more descriptive name and a docstring describing what the arguments are supposed to be.

Better might be to split it up into multiple smaller functions (each with docstrings, etc, of course!).

Also: I find a lot of the logic here quite hard to follow. I think that's partly because you've used separate lists of visits and fields and include the reference in the `visits` and `field` (should be plural) lists. It seems like this could be quite a bit simpler if you unified the visit and field for a given observation into a single data structure and then called `loadData` with a list of them. Simplify further by not including the reference frame in the list of targets. Something like this:
```python
obs = namedtuple("Observation", ["visit", "field"])
ref = obs(4192, 300)
targets = [obs(6377, 399)]
loadData(repo, ref, targets, camcol, filter)
```
You could then dispense with the special casing in the below (`if v == ref`, etc), and just iterate directly over the target list.

File: bin/check_astrometry.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+    Flags = ["base_PixelFlags_flag_saturated", "base_PixelFlags_flag_cr", "base_PixelFlags_flag_interpolated",
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: `Flags` should be `flags` with a lowercase `f`.

Also, is the aim here really just to eliminate all the flagged pixels? Rather than hard-coding the flag names, it might be safer to use something like `srcRef.getSchema()["base_PixelFlags"]`.

File: bin/check_astrometry.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+             "base_PsfFlux_flag_edge"]
+
+    #setup butler
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: Does this comment (and a few similar below) really help?

File: bin/check_astrometry.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        dataid = {'run':ref, 'filter':filter, 'field':ref_field, 'camcol':c}
+        oldSrc = butler.get('src', dataid, immediate=True)
+        print len(oldSrc), "sources in camcol :", c
+        if indx == 0 :
+            # retrieve the schema of the source catalog and extend it in order to add a field to record the camcol number
+            oldSchema = oldSrc.getSchema()
+            mapper = afwTable.SchemaMapper(oldSchema)
+            mapper.addMinimalSchema(oldSchema)
+            newSchema = mapper.getOutputSchema()
+            newSchema.addField("camcol", type=int, doc="camcol number")
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: I think you should prefer `mapper.addOutputField()` followed by `getOutputSchema()`. In practice, it'll probably make no difference for now, but I guess conceivably (a future version of) `SchemaMapper` could do some work to finalize mappings when `getOutputSchema()` is called that would be invalidated by further manipulation of the schema.

File: bin/check_astrometry.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+            continue
+        for indx, c in enumerate(camcol):
+            dataid = {'run':v, 'filter':filter, 'field':f, 'camcol':c}
+            if indx == 0 :
+                srcVis = butler.get('src', dataid, immediate=True)
+            else :
+                srcVis.extend(butler.get('src', dataid, immediate=True), False)
+            print len(srcVis), "sources in camcol : ", c
+        
+        match = afwTable.matchRaDec(srcRef, srcVis, afwGeom.Angle(1./3600., afwGeom.degrees))
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: Why not `afwGeom.Angle(1, afwGeom.arcseconds)`?

File: bin/check_astrometry.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+        matchNum = len(match)
+        print "Visit :", v, matchNum, "matches found"
+
+	schemaRef = srcRef.getSchema()
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: Indentation is wrong.

(Actually, I think the real problem here is that you're mixing tabs and spaces.)

Also, the manipulation of `srcRef` can all take place outside the enclosing loop.

File: bin/check_astrometry.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        schemaVis = srcVis.getSchema()
+        extRefKey = schemaRef["base_ClassificationExtendedness_value"].asKey()
+        extVisKey = schemaVis["base_ClassificationExtendedness_value"].asKey()
+        flagKeysRef = []
+        flagKeysVis = []
+        for f in Flags :
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: Isn't `f` the field? Can we use something else for `flag?

File: bin/check_astrometry.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+            dist.append(ang)
+
+    return pipeBase.Struct(
+        mag = mag,
+        dist = dist,
+        match = matchNum
+    )
+
+def check_astrometry(repo, mag, dist, match) :
+    # Plot angular distance between matched sources from different exposures
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: I comment below on the naming of this function -- is it plotting or checking?

Beyond that, do you actually want to generate the plot on each iteration? I'd suggest that the basic use case for this script will be to produce a simple pass/fail result; extra plotting or diagnostics should be optional.

File: bin/check_astrometry.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+    ax[1][1].set_ylim([0.,400.])
+    ax[1][0].set_xlabel("Distance in mas", fontsize=20)
+    ax[1][1].set_xlabel("Magnitude", fontsize=20)
+    ax[1][1].set_ylabel("Distance in mas", fontsize=20)
+    ax[1][0].tick_params(labelsize=20)
+    ax[1][1].tick_params(labelsize=20)
+
+    print "Median value of the astrometric scatter - all magnitudes:", np.median(dist), "mas"
+
+    good_mag_limit = 19.5
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: The definition of `good_mag_limit` is pretty well hidden inside this function, and it's not really justified. Can you make it more prominent and indicate why 19.5 is an appropriate value?

File: bin/check_astrometry.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+    ax[2][1].scatter(np.asarray(mag)[idxs], np.asarray(dist)[idxs], s=10, color='b')
+    ax[2][1].set_xlabel("Magnitude", fontsize=20)
+    ax[2][1].set_ylabel("Distance in mas - mag < %.1f" % good_mag_limit, fontsize=20)
+    ax[2][1].set_ylim([0.,200.])
+    ax[2][0].tick_params(labelsize=20)
+    ax[2][1].tick_params(labelsize=20)
+    
+    plt.suptitle("Astrometry Check", fontsize=30)
+    plotPath = os.path.join(repo, "astrometry_sdss.png")
+    plt.savefig(plotPath, format="png")
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: In practice, this is a throwaway repository, so it doesn't really matter. However, dumping arbitrary stuff into the repository rather than storing it through the Butler interface seems like a bad idea.

Even leaving that aside, unless the user actually knows that's where you're going to stash stuff, they'll probably never see your output. Why not just store to the working directory?

File: bin/check_astrometry.py [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+
+    astromScatter = np.median(np.asarray(dist)[idxs])
+    print "Astrometric scatter (median) - mag < %1.f :" % good_mag_limit, astromScatter, "mas"
+
+    return astromScatter
+    
+def main(repo):
+    
+    # List of SDSS runs to be considered
+    runs = [4192, 6377]
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: Would it be better to make all this configurable via the command line rather than hard-coding it in the script? (Perhaps even make this a `CmdLineTask`, and hence give it a standard interface.)

File: bin/check_astrometry.py [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+    field = [300, 399]
+
+    # Reference visit (the other viisits will be compared to this one
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: Typo ("viisits").

File: bin/check_astrometry.py [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+    
+    # Reference values for the median astrometric scatter and the number of matches
+    medianRef = 100
+    matchRef = 500
+    
+    struct = loadData(repo, runs, field, ref, ref_field, camcol, filter)
+    mag = struct.mag
+    dist = struct.dist
+    match = struct.match
+    astromScatter = check_astrometry(repo, mag, dist, match)
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: Check or plot? It's actually making a plot which you might check by eye _and_ returning a value which you check in code below. In neither case is this function call actually doing the checking.

File: bin/check_astrometry.py [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+
+    if astromScatter > medianRef :
+	print "Median astrometric scatter %.1f mas is larger than reference : %.1f mas "%(astromScatter, medianRef)
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: Line is too long.

Actually, in terms of number of characters, it's not. But you've indented with a tab, which should be four spaces. When you expand the tab, you're over the limit.

Also, please uses spaces around `%`.

File: bin/check_astrometry.py [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+	sys.exit(99)
```
Comment (jdswinbank) [inline bin/check_astrometry.py]: Why 99?

File: bin/check_astrometry.py (hunk 11 above)
Comment (PaulPrice) [inline bin/check_astrometry.py]: Why a camelCase name earlier and an underscore here?  What's the current policy?

File: bin/check_astrometry.py (hunk 8 above)
Comment (PaulPrice) [inline bin/check_astrometry.py]: Even better: `1.0*afwGeom.arcseconds`

File: bin/check_astrometry.py (hunk 6 above)
Comment (timj) [inline bin/check_astrometry.py]: I'll also note that there needs to be a space between the `#` and the text. Maybe run `flake8` on the file.

File: bin/check_astrometry.py (hunk 5 above)
Comment (PaulPrice) [inline bin/check_astrometry.py]: I think this is an example of something you're going to regret at some stage not making it easily configurable.

File: bin/check_astrometry.py (hunk 11 above)
Comment (jdswinbank) [inline bin/check_astrometry.py]: Coding standards call for camel case throughout.

Comment (timj) [inline bin/check_astrometry.py]: Although we need to update that. I think we agreed that new code could be PEP8 style so long as it's consistent within the module.
